gtsam/hybrid/HybridFactorGraph.cpp

```diff
@@ -0,0 +1,383 @@
+  for (auto &k : separatorKeys) {
+    if (discreteCardinalities.find(k) != discreteCardinalities.end()) {
+      discreteSeparatorSet.insert(discreteCardinalities.at(k));
+    } else {
+      continuousSeparator.insert(k);
+      allContinuousKeys.insert(k);
+    }
+  }
+
+  // Only for printing
+  if (DEBUG) {
+    std::cout << RED_BOLD << "Keys: " << RESET;
+    for (auto &f : frontalKeys) {
+      if (discreteCardinalities.find(f) != discreteCardinalities.end()) {
+        auto &key = discreteCardinalities.at(f);
+        std::cout << boost::format(" (%1%,%2%),") %
+                         DefaultKeyFormatter(key.first) % key.second;
+      } else {
+        std::cout << " " << DefaultKeyFormatter(f) << ",";
+      }
+    }
+
+    if (separatorKeys.size() > 0) {
+      std::cout << " | ";
+    }
+
+    for (auto &f : separatorKeys) {
+      if (discreteCardinalities.find(f) != discreteCardinalities.end()) {
+        auto &key = discreteCardinalities.at(f);
+        std::cout << boost::format(" (%1%,%2%),") %
+                         DefaultKeyFormatter(key.first) % key.second;
+      } else {
+        std::cout << DefaultKeyFormatter(f) << ",";
+      }
+    }
+    std::cout << "\n" << RESET;
+  }
+
+  // NOTE: We should really defer the product here because of pruning
+
+  // Case 1: we are only dealing with continuous
+  if (discreteCardinalities.empty() && !allContinuousKeys.empty()) {
+    if (DEBUG) {
+      std::cout << RED_BOLD << "CONT. ONLY" << RESET << "\n";
+    }
+
+    GaussianFactorGraph gfg;
+    for (auto &fp : factors) {
+      auto ptr = boost::dynamic_pointer_cast<HybridGaussianFactor>(fp);
+      if (ptr)
+        gfg.push_back(ptr->inner);
+      else
+        gfg.push_back(boost::static_pointer_cast<GaussianConditional>(
+            boost::static_pointer_cast<HybridConditional>(fp)->inner));
+    }
+
+    auto result = EliminatePreferCholesky(gfg, frontalKeys);
+    return std::make_pair(
+        boost::make_shared<HybridConditional>(result.first),
+        boost::make_shared<HybridGaussianFactor>(result.second));
+  }
+
+  // Case 2: we are only dealing with discrete
+  if (allContinuousKeys.empty()) {
+    if (DEBUG) {
+      std::cout << RED_BOLD << "DISCRETE ONLY" << RESET << "\n";
+    }
+
+    DiscreteFactorGraph dfg;
+    for (auto &fp : factors) {
+      auto ptr = boost::dynamic_pointer_cast<HybridDiscreteFactor>(fp);
+      if (ptr)
+        dfg.push_back(ptr->inner);
+      else
+        dfg.push_back(boost::static_pointer_cast<DiscreteConditional>(
+            boost::static_pointer_cast<HybridConditional>(fp)->inner));
+    }
+
+    auto result = EliminateDiscrete(dfg, frontalKeys);
+
+    return std::make_pair(
+        boost::make_shared<HybridConditional>(result.first),
+        boost::make_shared<HybridDiscreteFactor>(result.second));
+  }
+
+  // Case 3: We are now in the hybrid land!
+  // NOTE: since we use the special JunctionTree, only possiblity is cont.
+  // conditioned on disc.
```

Comment:
I don't understand this. Wouldn't hybrid and continuous only be similar because instead of a `GaussianMixtureFactor` we would just have a GaussianFactor?

Reply:
If you look at the Junction tree type for hybrid you will see that it checks the types of frontals so we never merge two cliques with both continuous and discrete frontals